### PR TITLE
Fixes #5440 Check for Action Scheduler missing tables

### LINKED ISSUE #5440: Improve detection of missing AS tables
**Before submitting an issue please check that you’ve completed the following steps:**
- Made sure you’re on the latest version
- Used the search feature to ensure that the bug hasn’t been reported before

**Describe the bug**
In the current approach we have a bug related to the not expected characters in the database name. Our query here is not using `` causing fail of the check and displaying false message:
https://github.com/wp-media/wp-rocket/blob/ba69c9365ac25fe9eb3c09d6910799bdfa13e225/inc/Engine/Preload/Activation/Activation.php#L117

**Expected behavior**
We should check the existence of the tables only in the specific conditions:
1. When activating the plugin, before Preload do any operation on the tables
2. When enabling RUCSS
3. When enabling Preload
4. During each update of the plugin

**Backlog Grooming (for WP Media dev team use only)**
- [ ] Reproduce the problem
- [ ] Identify the root cause
- [ ] Scope a solution
- [ ] Estimate the effort

### PR DESCRIPTION
**Pull request by remyperona**
## Description

Check for missing Action Scheduler tables when activating the plugin, updating, and enabling remove unused CSS or preload. Recreate them if they are missing.

Fixes #5440

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)

## Is the solution different from the one proposed during the grooming?

Create the check class in the health check namespace instead of the admin.

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Review by jeawhanlee — APPROVED** [review]

**Comment by Mai-Saad** [comment]
@Tabrisrp  Fatal error is there with the following scenario:
1- Using trunk while RUCSS is enabled
2- Deactivate WPR
3- Delete the actions table
4- Switch to current branch + composer dump autoload
5- Activate WPR
or 
1- Fresh install to branch
2- Drop actions table
3- Activate WPR
Can you please check?
```
[20-Sep-2022 08:24:08 UTC] PHP Fatal error:  Uncaught RuntimeException: Error saving action: Table 'new_rocketlabsqa_ovh_AzP4U3TX.wp_actionscheduler_actions' doesn't exist in /var/www/new.rocketlabsqa.ovh/htdocs/wp-content/plugins/wpforms-lite/vendor/woocommerce/action-scheduler/classes/data-stores/ActionScheduler_DBStore.php:86
Stack trace:
#0 /var/www/new.rocketlabsqa.ovh/htdocs/wp-content/plugins/wpforms-lite/vendor/woocommerce/action-scheduler/classes/ActionScheduler_ActionFactory.php(177): ActionScheduler_DBStore->save_action()
#1 /var/www/new.rocketlabsqa.ovh/htdocs/wp-content/plugins/wpforms-lite/vendor/woocommerce/action-scheduler/classes/ActionScheduler_ActionFactory.php(69): ActionScheduler_ActionFactory->store()
#2 /var/www/new.rocketlabsqa.ovh/htdocs/wp-content/plugins/wpforms-lite/vendor/woocommerce/action-scheduler/functions.php(19): ActionScheduler_ActionFactory->async()
#3 /var/www/new.rocketlabsqa.ovh/htdocs/wp-content/plugins/wp-rocket/inc/Engine/Common/Queue/AbstractASQueue.php(25): as_enqueue_async_action()
#4 /var/www/new.rocketlabsqa.ovh/htdocs/wp-content/plugins/wp-rocket/inc/Engine/Preload/Controller/Queue.php(24): WP_Rocket\Engine\Common\Queue\AbstractASQueue->add_async()
#5 /var/www/new.rocketlabsqa.ovh/htdocs/wp-content/plugins/wp-rocket/inc/Engine/Preload/Activation/Activation.php(45): WP_Rocket\Engine\Preload\Controller\Queue->add_job_preload_job_load_initial_sitemap_async()
#6 [internal function]: WP_Rocket\Engine\Preload\Activation\Activation->activate()
#7 /var/www/new.rocketlabsqa.ovh/htdocs/wp-content/plugins/wp-rocket/inc/Dependencies/League/Container/Inflector/Inflector.php(116): call_user_func_array()
#8 /var/www/new.rocketlabsqa.ovh/htdocs/wp-content/plugins/wp-rocket/inc/Dependencies/League/Container/Inflector/InflectorAggregate.php(53): WP_Rocket\Dependencies\League\Container\Inflector\Inflector->inflect()
#9 /var/www/new.rocketlabsqa.ovh/htdocs/wp-content/plugins/wp-rocket/inc/Dependencies/League/Container/Container.php(158): WP_Rocket\Dependencies\League\Container\Inflector\InflectorAggregate->inflect()
#10 /var/www/new.rocketlabsqa.ovh/htdocs/wp-content/plugins/wp-rocket/inc/Engine/Preload/Activation/ServiceProvider.php(57): WP_Rocket\Dependencies\League\Container\Container->get()
#11 /var/www/new.rocketlabsqa.ovh/htdocs/wp-content/plugins/wp-rocket/inc/Dependencies/League/Container/ServiceProvider/ServiceProviderAggregate.php(102): WP_Rocket\Engine\Preload\Activation\ServiceProvider->register()
#12 /var/www/new.rocketlabsqa.ovh/htdocs/wp-content/plugins/wp-rocket/inc/Dependencies/League/Container/Container.php(172): WP_Rocket\Dependencies\League\Container\ServiceProvider\ServiceProviderAggregate->register()
#13 /var/www/new.rocketlabsqa.ovh/htdocs/wp-content/plugins/wp-rocket/inc/Engine/Activation/Activation.php(47): WP_Rocket\Dependencies\League\Container\Container->get()
#14 /var/www/new.rocketlabsqa.ovh/htdocs/wp-includes/class-wp-hook.php(307): WP_Rocket\Engine\Activation\Activation::activate_plugin()
#15 /var/www/new.rocketlabsqa.ovh/htdocs/wp-includes/class-wp-hook.php(331): WP_Hook->apply_filters()
#16 /var/www/new.rocketlabsqa.ovh/htdocs/wp-includes/plugin.php(476): WP_Hook->do_action()
#17 /var/www/new.rocketlabsqa.ovh/htdocs/wp-admin/includes/plugin.php(691): do_action()
#18 /var/www/new.rocketlabsqa.ovh/htdocs/wp-admin/plugins.php(58): activate_plugin()
#19 {main}
  thrown in /var/www/new.rocketlabsqa.ovh/htdocs/wp-content/plugins/wpforms-lite/vendor/woocommerce/action-scheduler/classes/data-stores/ActionScheduler_DBStore.php on line 86
```
- if we deactivated WPForms lite then activate WPR , we will have this error 
```
[20-Sep-2022 08:32:22 UTC] PHP Fatal error:  Uncaught RuntimeException: Error saving action: Error saving action: Table 'new_rocketlabsqa_ovh_AzP4U3TX.wp_actionscheduler_actions' doesn't exist in /var/www/new.rocketlabsqa.ovh/htdocs/wp-content/plugins/wp-rocket/inc/Dependencies/ActionScheduler/classes/migration/ActionScheduler_DBStoreMigrator.php:44
Stack trace:
#0 /var/www/new.rocketlabsqa.ovh/htdocs/wp-content/plugins/wp-rocket/inc/Dependencies/ActionScheduler/classes/data-stores/ActionScheduler_HybridStore.php(242): ActionScheduler_DBStoreMigrator->save_action()
#1 /var/www/new.rocketlabsqa.ovh/htdocs/wp-content/plugins/wp-rocket/inc/Dependencies/ActionScheduler/classes/ActionScheduler_ActionFactory.php(177): ActionScheduler_HybridStore->save_action()
#2 /var/www/new.rocketlabsqa.ovh/htdocs/wp-content/plugins/wp-rocket/inc/Dependencies/ActionScheduler/classes/ActionScheduler_ActionFactory.php(69): ActionScheduler_ActionFactory->store()
#3 /var/www/new.rocketlabsqa.ovh/htdocs/wp-content/plugins/wp-rocket/inc/Dependencies/ActionScheduler/functions.php(19): ActionScheduler_ActionFactory->async()
#4 /var/www/new.rocketlabsqa.ovh/htdocs/wp-content/plugins/wp-rocket/inc/Engine/Common/Queue/AbstractASQueue.php(25): as_enqueue_async_action()
#5 /var/www/new.rocketlabsqa.ovh/htdocs/wp-content/plugins/wp-rocket/inc/Engine/Preload/Controller/Queue.php(24): WP_Rocket\Engine\Common\Queue\AbstractASQueue->add_async()
#6 /var/www/new.rocketlabsqa.ovh/htdocs/wp-content/plugins/wp-rocket/inc/Engine/Preload/Activation/Activation.php(45): WP_Rocket\Engine\Preload\Controller\Queue->add_job_preload_job_load_initial_sitemap_async()
#7 [internal function]: WP_Rocket\Engine\Preload\Activation\Activation->activate()
#8 /var/www/new.rocketlabsqa.ovh/htdocs/wp-content/plugins/wp-rocket/inc/Dependencies/League/Container/Inflector/Inflector.php(116): call_user_func_array()
#9 /var/www/new.rocketlabsqa.ovh/htdocs/wp-content/plugins/wp-rocket/inc/Dependencies/League/Container/Inflector/InflectorAggregate.php(53): WP_Rocket\Dependencies\League\Container\Inflector\Inflector->inflect()
#10 /var/www/new.rocketlabsqa.ovh/htdocs/wp-content/plugins/wp-rocket/inc/Dependencies/League/Container/Container.php(158): WP_Rocket\Dependencies\League\Container\Inflector\InflectorAggregate->inflect()
#11 /var/www/new.rocketlabsqa.ovh/htdocs/wp-content/plugins/wp-rocket/inc/Engine/Preload/Activation/ServiceProvider.php(57): WP_Rocket\Dependencies\League\Container\Container->get()
#12 /var/www/new.rocketlabsqa.ovh/htdocs/wp-content/plugins/wp-rocket/inc/Dependencies/League/Container/ServiceProvider/ServiceProviderAggregate.php(102): WP_Rocket\Engine\Preload\Activation\ServiceProvider->register()
#13 /var/www/new.rocketlabsqa.ovh/htdocs/wp-content/plugins/wp-rocket/inc/Dependencies/League/Container/Container.php(172): WP_Rocket\Dependencies\League\Container\ServiceProvider\ServiceProviderAggregate->register()
#14 /var/www/new.rocketlabsqa.ovh/htdocs/wp-content/plugins/wp-rocket/inc/Engine/Activation/Activation.php(47): WP_Rocket\Dependencies\League\Container\Container->get()
#15 /var/www/new.rocketlabsqa.ovh/htdocs/wp-includes/class-wp-hook.php(307): WP_Rocket\Engine\Activation\Activation::activate_plugin()
#16 /var/www/new.rocketlabsqa.ovh/htdocs/wp-includes/class-wp-hook.php(331): WP_Hook->apply_filters()
#17 /var/www/new.rocketlabsqa.ovh/htdocs/wp-includes/plugin.php(476): WP_Hook->do_action()
#18 /var/www/new.rocketlabsqa.ovh/htdocs/wp-admin/plugins.php(194): do_action()
#19 {main}
  thrown in /var/www/new.rocketlabsqa.ovh/htdocs/wp-content/plugins/wp-rocket/inc/Dependencies/ActionScheduler/classes/migration/ActionScheduler_DBStoreMigrator.php on line 44
```

**Comment by piotrbak** [comment]
We should handle the table recreation when activating the plugin 🤔

**Comment by jeawhanlee** [comment]
> We should handle the table recreation when activating the plugin 🤔

We are already doing that with the filter here but wierd, the callback is not fired at all and priority doesn't fix this as well in https://github.com/wp-media/wp-rocket/blob/2fdfe864015c4782504167d9d8e358080954ec00/inc/Engine/HealthCheck/ActionSchedulerCheck.php#L20
but calling the AS table recreation methods directly here https://github.com/wp-media/wp-rocket/blob/2fdfe864015c4782504167d9d8e358080954ec00/inc/Engine/Preload/Activation/Activation.php#L39 solves the issue but this doesn't seem consistent.

**Comment by remyperona** [comment]
Updated to handle the activation correctly

File: inc/Engine/Activation/ServiceProvider.php [hunk 1]
```diff
@@ -27,6 +28,7 @@ class ServiceProvider extends AbstractServiceProvider implements BootableService
 		'advanced_cache',
 		'capabilities_manager',
 		'wp_cache',
+		'actionscheduler_check'
```
Comment (jeawhanlee) [inline inc/Engine/Activation/ServiceProvider.php]:
```suggestion
		'actionscheduler_check',
```

**Review by jeawhanlee — CHANGES_REQUESTED** [review]

**Review by jeawhanlee — APPROVED** [review]

**Comment by Mai-Saad** [comment]
@Tabrisrp Working fine now when activating WPR or optimization option while the table is missing but have a fatal error when we do any of the following while the AS table is missing: 
- Clear cache and preload 
- Clear used CSS 
- Change the safelist of RUCSS 
- Switch theme while RUCSS/Preload are enabled
- Purge this URL while RUCSS/preload enabled
- Change permalinks while preload is enabled
- Change WPR settings (i.e enable Minify CSS and save)
- Add Category while preload/RUCSS enabled
```
[21-Sep-2022 07:55:40 UTC] WordPress database error Table 'new_rocketlabsqa_ovh_AzP4U3TX.wp_actionscheduler_actions' doesn't exist for query SELECT a.action_id FROM wp_actionscheduler_actions a WHERE 1=1 AND a.hook='rocket_preload_job_check_finished' AND a.status IN ('pending') ORDER BY a.scheduled_date_gmt ASC LIMIT 0, 5 made by do_action('admin_post_purge_cache'), WP_Hook->do_action, WP_Hook->apply_filters, do_admin_post_rocket_purge_cache, rocket_clean_domain, do_action('after_rocket_clean_domain'), WP_Hook->do_action, WP_Hook->apply_filters, WP_Rocket\Engine\Preload\Subscriber->clean_full_cache, WP_Rocket\Engine\Preload\Controller\Queue->add_job_preload_job_check_finished_async, WP_Rocket\Engine\Preload\Controller\Queue->job_preload_job_check_finished_async_exists, WP_Rocket\Engine\Common\Queue\AbstractASQueue->search, as_get_scheduled_actions, ActionScheduler_DBStore->query_actions
[21-Sep-2022 07:55:40 UTC] WordPress database error Table 'new_rocketlabsqa_ovh_AzP4U3TX.wp_actionscheduler_actions' doesn't exist for query 
INSERT INTO wp_actionscheduler_actions ( `hook`, `status`, `scheduled_date_gmt`, `scheduled_date_local`, `schedule`, `group_id`, `args` )
SELECT 'rocket_preload_job_check_finished', 'pending', '2022-09-21 07:56:40', '2022-09-21 07:56:40', 'O:30:\"ActionScheduler_SimpleSchedule\":2:{s:22:\"\0*\0scheduled_timestamp\";i:1663747000;s:41:\"\0ActionScheduler_SimpleSchedule\0timestamp\";i:1663747000;}', 1, '[1663746940]' FROM DUAL
WHERE ( SELECT NULL FROM DUAL ) IS NULL made by do_action('admin_post_purge_cache'), WP_Hook->do_action, WP_Hook->apply_filters, do_admin_post_rocket_purge_cache, rocket_clean_domain, do_action('after_rocket_clean_domain'), WP_Hook->do_action, WP_Hook->apply_filters, WP_Rocket\Engine\Preload\Subscriber->clean_full_cache, WP_Rocket\Engine\Preload\Controller\Queue->add_job_preload_job_check_finished_async, WP_Rocket\Engine\Common\Queue\AbstractASQueue->schedule_single, as_schedule_single_action, ActionScheduler_ActionFactory->single_unique, ActionScheduler_ActionFactory->store, ActionScheduler_DBStore->save_action, ActionScheduler_DBStore->save_action_to_db
[21-Sep-2022 07:55:40 UTC] PHP Fatal error:  Uncaught RuntimeException: Error saving action: Table 'new_rocketlabsqa_ovh_AzP4U3TX.wp_actionscheduler_actions' doesn't exist in /var/www/new.rocketlabsqa.ovh/htdocs/wp-content/plugins/action-scheduler/classes/data-stores/ActionScheduler_DBStore.php:118
Stack trace:
#0 /var/www/new.rocketlabsqa.ovh/htdocs/wp-content/plugins/action-scheduler/classes/data-stores/ActionScheduler_DBStore.php(62): ActionScheduler_DBStore->save_action_to_db()
#1 /var/www/new.rocketlabsqa.ovh/htdocs/wp-content/plugins/action-scheduler/classes/ActionScheduler_ActionFactory.php(247): ActionScheduler_DBStore->save_action()
#2 /var/www/new.rocketlabsqa.ovh/htdocs/wp-content/plugins/action-scheduler/classes/ActionScheduler_ActionFactory.php(117): ActionScheduler_ActionFactory->store()
#3 /var/www/new.rocketlabsqa.ovh/htdocs/wp-content/plugins/action-scheduler/functions.php(40): ActionScheduler_ActionFactory->single_unique()
#4 /var/www/new.rocketlabsqa.ovh/htdocs/wp-content/plugins/wp-rocket/inc/Engine/Common/Queue/AbstractASQueue.php(37): as_schedule_single_action()
#5 /var/www/new.rocketlabsqa.ovh/htdocs/wp-content/plugins/wp-rocket/inc/Engine/Preload/Controller/Queue.php(71): WP_Rocket\Engine\Common\Queue\AbstractASQueue->schedule_single()
#6 /var/www/new.rocketlabsqa.ovh/htdocs/wp-content/plugins/wp-rocket/inc/Engine/Preload/Subscriber.php(264): WP_Rocket\Engine\Preload\Controller\Queue->add_job_preload_job_check_finished_async()
#7 /var/www/new.rocketlabsqa.ovh/htdocs/wp-includes/class-wp-hook.php(309): WP_Rocket\Engine\Preload\Subscriber->clean_full_cache()
#8 /var/www/new.rocketlabsqa.ovh/htdocs/wp-includes/class-wp-hook.php(331): WP_Hook->apply_filters()
#9 /var/www/new.rocketlabsqa.ovh/htdocs/wp-includes/plugin.php(476): WP_Hook->do_action()
#10 /var/www/new.rocketlabsqa.ovh/htdocs/wp-content/plugins/wp-rocket/inc/functions/files.php(781): do_action()
#11 /var/www/new.rocketlabsqa.ovh/htdocs/wp-content/plugins/wp-rocket/inc/common/purge.php(438): rocket_clean_domain()
#12 /var/www/new.rocketlabsqa.ovh/htdocs/wp-includes/class-wp-hook.php(307): do_admin_post_rocket_purge_cache()
#13 /var/www/new.rocketlabsqa.ovh/htdocs/wp-includes/class-wp-hook.php(331): WP_Hook->apply_filters()
#14 /var/www/new.rocketlabsqa.ovh/htdocs/wp-includes/plugin.php(476): WP_Hook->do_action()
#15 /var/www/new.rocketlabsqa.ovh/htdocs/wp-admin/admin-post.php(85): do_action()
#16 {main}
  thrown in /var/www/new.rocketlabsqa.ovh/htdocs/wp-content/plugins/action-scheduler/classes/data-stores/ActionScheduler_DBStore.php on line 118
```

```
[21-Sep-2022 08:22:45 UTC] PHP Fatal error:  Uncaught RuntimeException: Error saving action: Error saving action: Table 'new_rocketlabsqa_ovh_AzP4U3TX.wp_actionscheduler_actions' doesn't exist in /var/www/new.rocketlabsqa.ovh/htdocs/wp-content/plugins/action-scheduler/classes/migration/ActionScheduler_DBStoreMigrator.php:44
Stack trace:
#0 /var/www/new.rocketlabsqa.ovh/htdocs/wp-content/plugins/action-scheduler/classes/data-stores/ActionScheduler_HybridStore.php(242): ActionScheduler_DBStoreMigrator->save_action()
#1 /var/www/new.rocketlabsqa.ovh/htdocs/wp-content/plugins/action-scheduler/classes/ActionScheduler_ActionFactory.php(247): ActionScheduler_HybridStore->save_action()
#2 /var/www/new.rocketlabsqa.ovh/htdocs/wp-content/plugins/action-scheduler/classes/ActionScheduler_ActionFactory.php(85): ActionScheduler_ActionFactory->store()
#3 /var/www/new.rocketlabsqa.ovh/htdocs/wp-content/plugins/action-scheduler/functions.php(22): ActionScheduler_ActionFactory->async_unique()
#4 /var/www/new.rocketlabsqa.ovh/htdocs/wp-content/plugins/wp-rocket/inc/Engine/Common/Queue/AbstractASQueue.php(25): as_enqueue_async_action()
#5 /var/www/new.rocketlabsqa.ovh/htdocs/wp-content/plugins/wp-rocket/inc/Engine/Optimization/RUCSS/Controller/Queue.php(71): WP_Rocket\Engine\Common\Queue\AbstractASQueue->add_async()
#6 /var/www/new.rocketlabsqa.ovh/htdocs/wp-content/plugins/wp-rocket/inc/Engine/Optimization/RUCSS/Controller/UsedCSS.php(578): WP_Rocket\Engine\Optimization\RUCSS\Controller\Queue->add_job_status_check_async()
#7 /var/www/new.rocketlabsqa.ovh/htdocs/wp-content/plugins/wp-rocket/inc/Engine/Optimization/RUCSS/Cron/Subscriber.php(94): WP_Rocket\Engine\Optimization\RUCSS\Controller\UsedCSS->process_pending_jobs()
#8 /var/www/new.rocketlabsqa.ovh/htdocs/wp-includes/class-wp-hook.php(307): WP_Rocket\Engine\Optimization\RUCSS\Cron\Subscriber->process_pending_jobs()
#9 /var/www/new.rocketlabsqa.ovh/htdocs/wp-includes/class-wp-hook.php(331): WP_Hook->apply_filters()
#10 /var/www/new.rocketlabsqa.ovh/htdocs/wp-includes/plugin.php(524): WP_Hook->do_action()
#11 /var/www/new.rocketlabsqa.ovh/htdocs/wp-cron.php(138): do_action_ref_array()
#12 {main}
  thrown in /var/www/new.rocketlabsqa.ovh/htdocs/wp-content/plugins/action-scheduler/classes/migration/ActionScheduler_DBStoreMigrator.php on line 44
```
- Clear this cache of a post will result in the following error
`[21-Sep-2022 08:31:25 UTC] WordPress database error Table 'new_rocketlabsqa_ovh_AzP4U3TX.wp_actionscheduler_actions' doesn't exist for query SELECT a.action_id FROM wp_actionscheduler_actions a LEFT JOIN wp_actionscheduler_groups g ON g.group_id=a.group_id WHERE 1=1 AND g.slug='rocket-rucss' AND a.status IN ('complete') AND a.last_attempt_gmt <= '2022-09-21 07:31:25' LIMIT 0, 100 made by do_action_ref_array('action_scheduler_run_queue_rucss'), WP_Hook->do_action, WP_Hook->apply_filters, WP_Rocket\Engine\Common\Queue\RUCSSQueueRunner->run, ActionScheduler_Abstract_QueueRunner->run_cleanup, ActionScheduler_QueueCleaner->clean, WP_Rocket\Engine\Common\Queue\Cleaner->delete_old_actions, ActionScheduler_HybridStore->query_actions, ActionScheduler_DBStore->query_actions`

**Comment by GeekPress** [comment]
Once the plugin will be updated to the new version, the issue must be fixed for all users. Maybe just a few will still get the error if the plugin doesn't have the permissions to write in the database. It will be a very rare case.

I don't think we need to cover all cases mentioned by Mai if the table is still missing.

**Comment by piotrbak** [comment]
Yes, this is what we agreed on the meeting.

**Comment by GeekPress** [comment]
Ok perfect. @Mai-Saad you can merge it to develop.

**Comment by Mai-Saad** [comment]
@wp-media/php Can anyone please resolve conflicts here?

**Comment by engahmeds3ed** [comment]
I'm Resolving them now

**Pull request merged**